src/object/srv_obj.c

```diff
@@ -1585,19 +1555,34 @@ obj_local_rw_internal(crt_rpc_t *rpc, struct obj_io_context *ioc, daos_iod_t *io
 
 	if (obj_rpc_is_fetch(rpc) && !spec_fetch &&
 	    daos_csummer_initialized(ioc->ioc_coc->sc_csummer)) {
-		rc = obj_fetch_csum_init(ioc->ioc_coc, orw, orwo);
+		/**
+		 * Allocate memory for the csum structures.
+		 * This memory and information will be used by VOS to put the checksums
+		 * in as it fetches the data's metadata from the btree/evtree.
+		 *
+		 * The memory will be freed in obj_rw_reply
+		 */
+		rc = daos_csummer_alloc_iods_csums(ioc->ioc_coc->sc_csummer, iods, iods_nr, false,
+						   NULL, &orwo->orw_iod_csums.ca_arrays);
```

Comment:
using iods/iods_nr here seems incorrect to me.
One thing is that for the case of one IO request (update or fetch) with multiple akeys/iods, if one akey/iod targets to multiple shards, and another akey/iod target for less shards or single shards, then the some akeys/iods possibly be removed - RPC's orw->orw_iod_array with all akeys/iods, but obj_get_iods_offs() possibly return less akeys/iods (the returned nr of obj_get_iods_offs possibly less than iod_array->oia_iod_nr), and obj_get_iods_offs() returns the "skips" bitmap to indicate which akey/iod in orw->orw_iod_array is removed/skipped.

But the "orwo->orw_iod_csums.ca_arrays" should include all the csum for all items in original RPC's orw->orw_iod_array so client can check with its original RPC request.
So if change like this way here, seems incorrect to me.

But I am not sure what caused original csum issue this patch try to address.

Reply:
I see, thanks!
The issue I'm trying to solve is when there are multiple iods/akeys in a single IO request and each iod's size is DAOS_REC_ANY. The piece that's missing for checksums when iods != orw->orw_iod_array.oia_iods is the iod's size. A side effect of the vos_fetch_update call in obj_local_rw_internal is that it sets the iods' sizes.
I guess I can just copy the size from iods to orw->orw_iod_array.oia_iods. I'll work on it some more.